elfloader-tool/src/arch-riscv/crt0.S

```diff
@@ -30,64 +50,93 @@ _start:
   addi  gp, gp, %pcrel_lo(1b)
 .option pop
 
-/* a0 should have hart id, store it in s0 so as not to clobber from HSM calls
-   a1 should have the dtb, store it in s2 so as not to clobber it as well */
-  mv s0, a0
-  mv s2, a1
+  /* save the parameters passed */
+  mv s0, a0 /* preserve a0 (hart id) in s0 */
+  mv s1, a1 /* preserve a1 (dtb) in s1 */
 
 #ifdef CONFIG_IMAGE_BINARY
 /* Clear the BSS before we get to do anything more specific */
   jal clear_bss
 #endif
 
-/* Check Heart State Management extension exists so we can use that to switch harts
-   if we are not hart CONFIG_FIRST_HART_ID, otherwise rely on the legacy extension. */
-
-/* sbi_probe_extension: Returns 0 if the given SBI extension ID (EID) is not available,
-   or an extension-specific non-zero value if it is available.*/
-
-  jal spi_probe_extension
-  mv a0, s0
-  li s1, 1
-  blt x1, s1, _start1
-
-/* Update bool to tell boot code HSM exists */
-  la t1, hsm_exists
-  li t2, 1
-  amoadd.w t1, t2, (t1)
-
-  mv a0, s0
-  li s1, CONFIG_FIRST_HART_ID
-  beq  a0, s1, _start1
-
-/* start hart CONFIG_FIRST_HART_ID */
-hsm_switch_hart:
+  /* Check if the Heart State Management (HSM) extension exists, so it can be
+   * used to switch harts if we are not running on hart CONFIG_FIRST_HART_ID.
+   * The SBI returns SBI_SUCCESS (0) in a0 if the call could be processed or an
+   * error code in not. On SBI_SUCCESS the value in a1 is 0 is the extension is
+   * not available or an extension-specific non-zero value if it is available.
+   */
+  li a7, SBI_EXT_BASE
+  li a6, SBI_EXT_BASE_PROBE_EXT
+  li a0, SBI_HSM_BASE
+  ecall /* call SBI to probe for HSM extension */
+  seqz a0, a0  /* a0 = (SBI_SUCCESS == a0) ? 1 : 0 */
+  beqz a0, update_hsm_exists /* not SBI_SUCCESS -> set hsm_exists = 0 */
+  snez a0, a1  /* a0 = (0 != a1) ? 1 : 0 */
+update_hsm_exists:
+  /* Update global bool variable to tell boot code the HSM extension exists. */
+  la t0, hsm_exists
+  sw a0, 0(t0) /* 32-bit store operation: [t0] = (int32)a0 */
+
+  /* Check if we are running on CONFIG_FIRST_HART_ID. */
+  li a1, CONFIG_FIRST_HART_ID
+  bne  a1, s0, not_first_hart  /* if not FIRST_HART then -> not_first_hart */
+prepare_params_for_main:
+  mv a0, s0 /* restore hart id */
+  mv a1, s1 /* restore dtb */
+start_main:
+  la sp, (elfloader_stack_alloc + BIT(12))
+  /* The C code expects the registers to be set up as:
+   *   a0 = hart id
+   *   a1 = dtb
+   */
+  la ra, main
+  jr ra
+
+not_first_hart:
+  /* Seems we are not on FIRST_HART. If the HSM SBI extension is missing, there
+   * is nothing we can do here to switch harts. Continue boot, maybe the C code
+   * can handle this case somehow via platform proprietary ways, otherwise it
+   * should log a message and halt.
+   */
+  beq  a0, x0, prepare_params_for_main
```

Comment:
use `beqz a0, prepare_params_for_main` to avoif explicit x0 refernece, see https://github.com/riscv-non-isa/riscv-asm-manual/blob/master/riscv-asm.md#a-listing-of-standard-risc-v-pseudoinstructions